[Docs] PR for Document Inspect & Modify flow (#12724)
yugabyte/yugabyte-db: docs/content/preview/deploy/multi-dc/async-replication.md

  000030a5000030008000000000004000,000030a5000030008000000000004005,dfef757c415c4b2cacc9315b8acb539a
```

> When universes use different certificates, you need to store the certificates for the producer universe on the target universe, as follows:

Author: fixed

1. Ensure that `use_node_to_node_encryption` is set to `true` on all [masters](../../reference/configuration/yb-master/#use-node-to-node-encryption) and [t-servers](../../reference/configuration/yb-tserver/#use-node-to-node-encryption) on both the source and target.

1. For each master and t-server on the target universe, set the gflag `certs_for_cdc_dir` to the parent directory where you want to store all the source universe's certificates for replication.

1. Find the certificate authority file used by the source universe (`ca.crt`). This should be stored within the [`--certs_dir`]`/preview/reference/configuration/yb-master/#certs-dir`.

1. Copy this file to each node on the target. It needs to be copied to a directory named`<certs_for_cdc_dir>/<source_universe_uuid>`.

>     For example, if you previously set `certs_for_cdc_dir=/home/yugabyte/yugabyte_producer_certs`, and the source universe's ID is `00000000-1111-2222-3333-444444444444`, then you would need to copy the certificate file to `/home/yugabyte/yugabyte_producer_certs/00000000-1111-2222-3333-444444444444/ca.crt`.

Author: fixed

    ```

2. Take the backup of the tables on the source universe and restore at the target universe by following instructions from [Backup and restore](../../../manage/backup-restore/).
3. Execute the following command to set up the replication stream using the bootstrap IDs generated in step 1. Ensure that the bootstrap IDs are in the same order as their corresponding table IDs.

      ```sh
    ./bin/yb-admin -master_addresses <target_universe_master_addresses> setup_universe_replication \
      <source_universe_uuid>_<replication_stream_name> <source_universe_master_addresses> \
      <comma_separated_source_universe_table_ids> <comma_separated_bootstrap_ids>
    ```

    Consider the following example:

      ```sh
    ./bin/yb-admin-master_addresses 127.0.0.11:7100,127.0.0.12:7100,127.0.0.13:7100 setup_universe_replication \
      00000000-1111-2222-3333-444444444444_xCluster1 127.0.0.1:7100,127.0.0.2:7100,127.0.0.3:7100 \
      000033e1000030008000000000004000,000033e1000030008000000000004003,000033e1000030008000000000004006 \
      fb156717174941008e54fa958e613c10,a2a46f5cbf8446a3a5099b5ceeaac28b,c967967523eb4e03bcc201bb464e0679
      ```

> You can modify the bootstrap as follows:

Reviewer: Should this be a subitem in the page outline?

Author: Added a section header

- To wipe the test setup, use the `delete_universe_replication` command.

> - After running the `bootstrap_cdc_producer` command, you can verify entries via the `list_cdc_streams` command.

Author: Changed

You can also perform the following modifications:

- To add a table to the source and target universes, use the `alter_universe_replication add_table` command.
- To remove an existing table from the source and target universes, use the `alter_universe_replication remove_table` command.
- To change the master nodes on the source universe, execute the `alter_universe_replication set_master_addresses` command.
- You can verify changes via the `get_universe_config` command.

## Migrate schema

You can execute DDL operations after replication has been already configured for some tables.

### Stop user writes

In certain cases, it is possible to temporarily stop incoming user writes. You would typically approach this as follows:
- Stop any new incoming user writes.
- Wait for all changes to get replicated to the target universe. This can be observed by replication lag dropping to 0.
- Apply the DDL changes on both universes and [alter replication](../../../admin/yb-admin/#alter-universe-replication) for any newly-created tables. For example, after executing the `CREATE TABLE` or `CREATE INDEX` statements.
- Resume user writes.

### Use backup and restore

> If you cannot stop incoming user traffic, then the recommended approach would be to apply DDLs on the source universe combined with bootstrapping. Specifically, you would need to do the following:

Author: Changed

      000030a5000030008000000000004000,000030a5000030008000000000004005,dfef757c415c4b2cacc9315b8acb539a
    ```

## Set up replication with geo-partitioning

You start by creating a source and a target universe with the same configurations (the same regions and zones), as follows:

* Regions: EU(Paris), Asia Pacific(Mumbai) and US West(Oregon)

> * Zones: eu-west-3a, ap-south-1a and us-west-2a

Author: fixed

  ```sh
  ./bin/yb-ctl --rf 3 create --placement_info "cloud1.region1.zone1,cloud2.region2.zone2,cloud3.region3.zone3"
  ```

  Consider the following example:

  ```sh
  ./bin/yb-ctl --rf 3 create --placement_info "aws.us-west-2.us-west-2a,aws.ap-south-1.ap-south-1a,aws.eu-west-3.eu-west-3a"
  ```

> Create tables, table spaces, and partition tables at both the source and target universes, as per the following example:

Author: fixed

* Main table: transactions

> * Table spaces: eu_ts, ap_ts and us_ts

Author: fixed

                    FOR VALUES IN ('EU') TABLESPACE eu_ts;
  
  CREATE TABLE transactions_in
                    PARTITION OF transactions
                    (user_id, account_id, geo_partition, amount, created_at,
                    PRIMARY KEY (user_id HASH, account_id, geo_partition))
                    FOR VALUES IN ('IN') TABLESPACE ap_ts;
  
  CREATE TABLE transactions_us
                    PARTITION OF transactions
                    (user_id, account_id, geo_partition, amount, created_at,
                    PRIMARY KEY (user_id HASH, account_id, geo_partition))
                    DEFAULT TABLESPACE us_ts;
  ```

To create unidirectional replication, peform the following:

1. Collect partition table UUIDs from the source universe (partition tables, transactions_eu, transactions_in, transactions_us) by navigating to **Tables** in the Admin UI available at 127.0.0.1:7000. These UUIDs are to be used while setting up replication. <br><br>

  ![xCluster_with_GP](/images/explore/yb_xcluster_table_uuids.png)

<br>

2. Run the replication setup command for the source universe, as follows:
    ```sh
    ./bin/yb-admin -master_addresses <consumer_master_addresses> \

>     setup_universe_replication <producer universe UUID>_<replication_stream_name> \

Author: fixed

    <producer_master_addresses> <comma_separated_table_ids>
    ```

    Consider the following example:
    ```sh
    ./bin/yb-admin -master_addresses 127.0.0.11:7100,127.0.0.12:7100,127.0.0.13:7100 \
    setup_universe_replication 00000000-1111-2222-3333-444444444444_xClusterSetup1 \
    127.0.0.1:7100,127.0.0.2:7100,127.0.0.3:7100 \
    000033e1000030008000000000004007,000033e100003000800000000000400d,000033e1000030008000000000004013
    ```
3. Optionally, if you have access to YugabyteDB Anywhere, you can observe the replication setup (`xClusterSetup1`) by navigating to **Replication** on the source and target universe.

## Set up asynchronous replication in Kubernetes

> In the Kubernetes environment, you can set up a pod to pod connectivity, as follows:

Author: It was connectivity here (the section looked like a developer wrote it)

- Create a source and a target universe.

- Create tables in both universes, as follows: 

>   - Execute the following commands at for the source universe:

Reviewer 2: commands at for the

Author: fixed

      ```sh
    kubectl exec -it -n <source_universe_namespace> -t <source_universe_master_leader> -c <source_universe_container> -- bash
      /home/yugabyte/bin/ysqlsh -h <source_universe_yqlserver>
      create table query
      ```

    Consider the following example:

      ```sh
    kubectl exec -it -n xcluster-source -t yb-master-2 -c yb-master -- bash
      /home/yugabyte/bin/ysqlsh -h yb-tserver-1.yb-tservers.xcluster-source
      create table employees(id int primary key, name text);
      ```

>   - Execute the following commands at for the target universe:

Reviewer 2: commands at for the

Author: fixed

- Set up replication from the source universe by executing the following command on the source universe:

    ```sh
  kubectl exec -it -n <source_universe_namespace> -t <source_universe_master_leader> -c \
    <source_universe_container> -- bash -c "/home/yugabyte/bin/yb-admin -master_addresses \
    <target_universe_master_addresses> setup_universe_replication \
    <source_universe_UUID>_<replication_stream_name> <source_universe_master_addresses> \
    <comma_separated_table_ids>"
    ```

  Consider the following example:

    ```sh
  kubectl exec -it -n xcluster-source -t yb-master-2 -c yb-master -- bash -c \
    "/home/yugabyte/bin/yb-admin -master_addresses yb-master-2.yb-masters.xcluster-target.svc.cluster.local, \
    yb-master-1.yb-masters.xcluster-target.svc.cluster.local,yb-master-0.yb-masters.xcluster-target.svc.cluster.local \
    setup_universe_replication ac39666d-c183-45d3-945a-475452deac9f_xCluster_1 \
    yb-master-2.yb-masters.xcluster-source.svc.cluster.local,yb-master-1.yb-masters.xcluster-source.svc.cluster.local, \
    yb-master-0.yb-masters.xcluster-source.svc.cluster.local 00004000000030008000000000004001"
    ```

> - Perform the following on the source universe and then observe replication one the target universe:

Reviewer 2: ```suggestion
- Perform the following on the source universe and then observe replication on the target universe:
```

      ```sh
    ./bin/yb-admin-master_addresses 127.0.0.11:7100,127.0.0.12:7100,127.0.0.13:7100 setup_universe_replication \
      00000000-1111-2222-3333-444444444444_xCluster1 127.0.0.1:7100,127.0.0.2:7100,127.0.0.3:7100 \
      000033e1000030008000000000004000,000033e1000030008000000000004003,000033e1000030008000000000004006 \
      fb156717174941008e54fa958e613c10,a2a46f5cbf8446a3a5099b5ceeaac28b,c967967523eb4e03bcc201bb464e0679
      ```

You can modify the bootstrap as follows:

- To wipe the test setup, use the `delete_universe_replication` command.
- After running the `bootstrap_cdc_producer` command, you can verify entries via the `list_cdc_streams` command.

You can also perform the following modifications:

- To add a table to the source and target universes, use the `alter_universe_replication add_table` command.
- To remove an existing table from the source and target universes, use the `alter_universe_replication remove_table` command.
- To change the master nodes on the source universe, execute the `alter_universe_replication set_master_addresses` command.
- You can verify changes via the `get_universe_config` command.

## Migrate schema

> You can execute DDL operations, after replication has been already configured for some tables.

Reviewer 2: You can execute DDL operations, even after replication has been already configured for some tables.
or
You can execute DDL operations after replication has been already configured for some tables.
?

### Stop user writes

In certain cases, it is possible to temporarily stop incoming user writes. You would typically approach this as follows:
- Stop any new incoming user writes.
- Wait for all changes to get replicated to the target universe. This can be observed by replication lag dropping to 0.
- Apply the DDL changes on both universes and [alter replication](../../../admin/yb-admin/#alter-universe-replication) for any newly-created tables. For example, after executing the `CREATE TABLE` or `CREATE INDEX` statements.
- Resume user writes.

### Use backup and restore

> If you cannot stop incoming user traffic, then the recommended approach would be to apply DDLs on the source universe combined with bootstrapping. Specifically, you would need to do the following:

Reviewer 2: ```suggestion
If you can't stop incoming user traffic, then the recommended approach would be to apply DDLs on the source universe combined with bootstrapping. Specifically, you would need to do the following:
```

- Stop replication before making any DDL changes.
- Apply all your DDL changes to the source universe.

> - Backup the source universe and all the relevant tables for which you intend to replicate changes. Follow instructions provided in [Bootstrap a target universe](#bootstrap-a-target-universe).

Reviewer 2: ```suggestion
- Back up the source universe and all the relevant tables for which you intend to replicate changes. Follow instructions provided in [Bootstrap a target universe](#bootstrap-a-target-universe).
```

Author: fixed
